src/crypto/tls/common.go

```diff
@@ -278,10 +293,31 @@ type ConnectionState struct {
 	// RFC 7627, and https://mitls.org/pages/attacks/3SHAKE#channelbindings.
 	TLSUnique []byte
 
+	// ECHStatus reports the status of this connection's usage of the ECH
+	// extension.
+	ECHStatus ECHStatus
+
 	// ekm is a closure exposed via ExportKeyingMaterial.
 	ekm func(label string, context []byte, length int) ([]byte, error)
 }
 
+// ECHStatus represents a connection's ECH status.
+type ECHStatus struct {
+	// Offered indicates whether the client offered the ECH extension. Note that
+	// this reports real ECH usage only and does not reflect whether the client
+	// offered a dummy ECH extension.
+	Offered bool
+
+	// Accepted indicates whether the client-facing server accepted ECH for this
+	// connection.
+	Accepted bool
+
+	// RetryConfigs encodes the sequence of ECH configurations with which the
+	// client can retry the connection. This is set only if the client-facing
+	// server rejects ECH.
+	RetryConfigs []byte
```

Comment:
Question: should this value be exposed here? If so, should it be `[]byte`, `[]ECHConfig`, or some other opaque type? It should only be used once for a subsequent retry connection according to the spec, it should not replace the value obtained from, say, DNS. If `[]byte` or `[]ECHConfig` are in use, then the application developer could misuse it.

https://tlswg.org/draft-ietf-tls-esni/draft-ietf-tls-esni.html#section-6.3.2
> The value may only be applied to the retry connection. The client MUST continue to use the previously-advertised keys for subsequent connections.

Reply:
Removed ECHStatus.RetryConfigs.

Good point, I forgot about this rule. Personally I don't see a reason for it. In any case, the answer is that the retry 
configs should not be exposed in the public API, if the client is fully compliant. However, our client is **not** compliant, since it does not retry after rejection: it simply leaves the connection open. (This is useful for testing purposes.) For this reason it seemed like a good idea to put them in the public API so that the correct behavior could be implemented by the caller. However, since we removed `AbortIfRetryRequired()`, this isn't possible anymore.